app/models/sources/joyent.rb

```diff
@@ -0,0 +1,70 @@
+module Sources
+  class Joyent < Sources::Fog
+    class << self
+      def required_fields
+        super.merge({
+          username: :string,
+          password: :password,
+          datacenter: :string,
+          image_id: :string,
+          package_id: :string
+        })
+      end
+    end
+
+    # Connect to Joyent to make sure our config is valid
+    # You cannot verify custom images through Fog, but an incorrect image
+    # UUID will cause an error when creating the server.
+    def validate_config!
+      if !config['image_id'].nil?
+        connection.flavors.each do |f|
+          if f.name == config['package_id']
+            return true
+          end
+        end
+      end
+      return false
+    end
+
+    private
+
+    # server_is_proxy_type?(server)
+    # given a server, determine if it is running a proxy
+    def server_is_proxy_type?(server)
+      return server.image == config['image_id'] \
```

Comment:
What about datacenter?  The current production Zartan has 3 DigitalOcean Source objects each in a different region.  Without a datacenter check then different Joyant Source objects can pick up each other's created servers.